From: KOSAKI Motohiro <kosaki.motohiro@jp.fujitsu.com>
Subject: mm, mem-hotplug: recalculate lowmem_reserve when memory hotplug occurs
References: fate#311831,bnc#703100
Patch-Mainline: v3.0-rc5
Git-commit: f957db4fcdd8f03e186aa8f041f4049e76ab741c

Signed-off-by: Thomas Renninger <trenn@suse.de>
    
Currently, memory hotplug calls setup_per_zone_wmarks() and
calculate_zone_inactive_ratio(), but doesn't call
setup_per_zone_lowmem_reserve().

It means the number of reserved pages aren't updated even if memory hot
plug occur.  This patch fixes it.

Signed-off-by: KOSAKI Motohiro <kosaki.motohiro@jp.fujitsu.com>
Reviewed-by: KAMEZAWA Hiroyuki <kamezawa.hiroyu@jp.fujitsu.com>
Acked-by: Mel Gorman <mel@csn.ul.ie>
Reviewed-by: Minchan Kim <minchan.kim@gmail.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

---
 include/linux/mm.h  |    2 +-
 mm/memory_hotplug.c |    8 ++++----
 mm/page_alloc.c     |    4 ++--
 3 files changed, 7 insertions(+), 7 deletions(-)

Index: linux-2.6.32-SLE11-SP2/include/linux/mm.h
===================================================================
--- linux-2.6.32-SLE11-SP2.orig/include/linux/mm.h
+++ linux-2.6.32-SLE11-SP2/include/linux/mm.h
@@ -1328,7 +1328,7 @@ extern void set_dma_reserve(unsigned lon
 extern void memmap_init_zone(unsigned long, int, unsigned long,
 				unsigned long, enum memmap_context);
 extern void setup_per_zone_wmarks(void);
-extern void calculate_zone_inactive_ratio(struct zone *zone);
+extern int __meminit init_per_zone_wmark_min(void);
 extern void mem_init(void);
 extern void __init mmap_init(void);
 extern void show_mem(void);
Index: linux-2.6.32-SLE11-SP2/mm/memory_hotplug.c
===================================================================
--- linux-2.6.32-SLE11-SP2.orig/mm/memory_hotplug.c
+++ linux-2.6.32-SLE11-SP2/mm/memory_hotplug.c
@@ -457,8 +457,9 @@ int online_pages(unsigned long pfn, unsi
 		zone_pcp_update(zone);
 
 	mutex_unlock(&zonelists_mutex);
-	setup_per_zone_wmarks();
-	calculate_zone_inactive_ratio(zone);
+
+	init_per_zone_wmark_min();
+
 	if (onlined_pages) {
 		kswapd_run(zone_to_nid(zone));
 		node_set_state(zone_to_nid(zone), N_HIGH_MEMORY);
@@ -906,8 +907,7 @@ repeat:
 	zone->zone_pgdat->node_present_pages -= offlined_pages;
 	totalram_pages -= offlined_pages;
 
-	setup_per_zone_wmarks();
-	calculate_zone_inactive_ratio(zone);
+	init_per_zone_wmark_min();
 
 	vm_total_pages = nr_free_pagecache_pages();
 	writeback_set_ratelimit();
Index: linux-2.6.32-SLE11-SP2/mm/page_alloc.c
===================================================================
--- linux-2.6.32-SLE11-SP2.orig/mm/page_alloc.c
+++ linux-2.6.32-SLE11-SP2/mm/page_alloc.c
@@ -5003,7 +5003,7 @@ static void __setup_per_zone_wmarks(void
  *    1TB     101        10GB
  *   10TB     320        32GB
  */
-void calculate_zone_inactive_ratio(struct zone *zone)
+static void __meminit calculate_zone_inactive_ratio(struct zone *zone)
 {
 	unsigned int gb, ratio;
 
@@ -5115,7 +5115,7 @@ EXPORT_SYMBOL_GPL(adjust_memalloc_reserv
  * 8192MB:	11584k
  * 16384MB:	16384k
  */
-static int __init init_per_zone_wmark_min(void)
+int __meminit init_per_zone_wmark_min(void)
 {
 	unsigned long lowmem_kbytes;
 
